src/cycle_basis.c

```diff
@@ -0,0 +1,462 @@
+ *
+ * A review:
+ * https://www.sciencedirect.com/science/article/pii/S1574013709000483?casa_token=B7FmnsCBElwAAAAA:bfOYZDhzHpP24ym_2S2n5VuEU-dYsdAQJdmH3nvk8efkFrVmJkEY-KhHXawLZfoMFAKPIdHE
+ *
+ * The algorithm implemented here:
+ * https://dl.acm.org/doi/abs/10.1145/1644015.1644023
+ * 
+ */
+
+
+int igraph_minimum_cycle_basis(const igraph_t *graph,
+                const igraph_vector_t *weights,
+                igraph_vector_ptr_t *basis,
+		) {
+
+    /* General plan:
+     *
+     * 0. Find a feedback vertex set Z: all cycles pass through a vertex
+     *    in Z. Worst case Z := {all vertices}. Basically, 2-degree nodes
+     *    are excluded because there's no decision: the path goes in one
+     *    way and out another. Of course, 0-degree are also not needed.
+     * 1. Construct all shortest path trees 
+     * 2. Construct candidate set of cycle bases, sorted in nondecreasing
+     *    order of weight
+     * 3. for i = 1 to N do
+     *     3.1 compute vector Si s.t. <Cj, Si> = 0 for each 1 <= j < i
+     *         (orthogonal to the current cycles)
+     *     3.2 for all trees, update vertex labels based on Si
+     *     3.3 for each candidate cycle, compute <C, Si> using the labels
+     *     3.4 append the least-weight cycle non-orthogonal to Si
+     *         (so we can normalize it into <C, Si> = 1
+     *
+     * */
+    /* TODO: implement the actual functions */
+
+    long int i;
+    long int no_of_nodes = igraph_vcount(graph);
+    igraph_vector_t si;
+    igraph_vector_int_t nonorth_cycles;
+
+    igraph_vector_t fvs, basis_weight;
+    /* list of vertices = tree -> list of trees */
+    igraph_vector_ptr_t trees_z;
+    /* list of vertices = vector -> list of vectors = basis -> list of bases */
+    igraph_vector_ptr_t candidate_cycles;
+    /* total weight of each basis (?) */
+    igraph_vector_t candidate_weights;
+
+    /* 0. Feedback vertex set */
+    IGRAPH_CHECK(igraph_i_feedback_vertex_set_approx(
+			    graph, weights, &fvs));
+
+    /* 1. Construct all shortest path trees */
+    IGRAPH_CHECK(igraph_i_shortest_path_trees(
+			    graph, weights, &fvs, &trees_z));
+
+    /* 2. Construct candidate list */
+    IGRAPH_CHECK(igraph_i_candidate_bases(
+			    graph, weights,
+			    &candidate_cycles, &candidate_weights));
+
+    /* 3. Construct basis one by one */
+    for (i = 0; i < no_of_nodes; i++) {
+        /* 3.1 compute Si */
+	igraph_i_compute_Si(graph, weights, basis, i, &si);
+
+	/* 3.2 update tree vertex labels from Si */
+	igraph_i_update_tree_z_vertex_labels(
+			&trees_z, &si);
+
+	/* 3.3 find non-orthogonal candidates */
+	igraph_i_nonorthogonal_candidates(
+			candidate_cycles, &si, &nonorth_cycles);
+
+	/* 3.4 get the shortest nonorthogonal cycle */
+	igraph_i_shortest_nonorthogonal_cycle(
+			candidate_cycles, &candidate_weights, &nonorth_cycles, basis);
+
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_i_feedback_vertex_set_approx(const igraph_t *graph,
```

Comment:
`igraph_feedback_arc_set()` already contains an `algorithm` argument that opens up the possibility of implementing different approximation heuristics right within `igraph_feedback_arc_set()`. I propose adding whatever you plan to implement to `igraph_feedback_arc_set()` instead.